guide.md

```diff
@@ -0,0 +1,324 @@
+`--config` flag is provided, like so:
+
+```
+build:nix --host_platform=@io_tweag_rules_nixpkgs//nixpkgs/platforms:host
+run:nix   --host_platform=@io_tweag_rules_nixpkgs//nixpkgs/platforms:host
+build:nix --crosstool_top=@nixpkgs_config_cc//:toolchain
+```
+
+None of these options will apply to `bazel build` or `bazel run` unless an
+extra `--config nix` parameter is passed:
+
+```
+> bazel run //src:hello-world --config=nix
+...
+```
+
+# Working from a Nix-provided developer environment
+
+Another practical advantage of using Nix is that we can provide consistent,
+zero-setup development environments to all contributors to a project. These
+function similarly to, say, Python virtualenvs, but they're not just limited to
+providing dependencies for just one language ecosystem. Even better, they can
+provide system dependencies too.
+
+```
+> hello
+hello: command not found
+> nix-shell -p hello
+nix-shell> hello
+Hello, world!
+nix-shell> exit
+> hello
+hello: command not found
+```
+
+There's still a few loose ends in our setup that Nix developer environments can
+help tie up. First, the version of Bazel we've been using so far, has been
+unspecified. The general practice is to simply use one's distribution's package
+manager, something that doesn't deliver consistent Bazel versions across major
+OS platforms. Second, we haven't _disabled_ Bazel's deep-rooted tendency to 
+depend on the global environment. Indeed, you can rename `.bazelrc` and rerun
+`bazel run`, and everything will work just fine: Bazel will simply fall back to
+`@local_config_cc`, because a C++ compiler toolchain is still globally available.
+
+We can provide a Nix developer environment that only provides the specific dependencies that developers need:
+
+```nix
+# flake.nix
+{
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/nixos-22.05";
+    flake-compat = {
+      url = github:edolstra/flake-compat;
+      flake = false;
+    };
+    flake-utils.url = github:numtide/flake-utils;
+  };
+  outputs = { nixpkgs, flake-utils, ... }:
+    flake-utils.lib.eachDefaultSystem (system:
+      let pkgs = import nixpkgs { inherit system; };
+      in
+      {
+        devShells.default = with pkgs; mkShellNoCC {
+          name = "rules_nixpkgs_shell";
+          packages = [ bazel_5 bazel-buildtools cacert gcc nix git ];
+        };
+      });
+}
+```
+
+Now developers on different platforms can enter a development environment by
+running `nix develop`. We've also removed the "default" system-wide CC
+toolchain (using `mkShellNoCC` instead of the more common `mkShell`), so
+`@local_config_cc` will no longer work:
+
+```
+> bazel build //src:hello-world 
+INFO: Build options --crosstool_top, --host_crosstool_top, --host_platform, and 1 more have changed, discarding analysis cache.
+[snip]
+Auto-Configuration Error: Cannot find gcc or CC; either correct your path or set the CC environment variable
+```
```

Comment:
The guide should make it very clear when a snippet or described approach is considered "correct" and when it is "wrong". I'd also err on the side of "correct" first and "wrong" after (if at all). The reasoning here is this: Guides are often not read top-to-bottom carefully, but rather jumped around in or taken as reference. So, it should be very clear whether a given snippet or approach is safe to copy or not. E.g. the Rust book does this nicely with big visual indications.

Reply:
I also think you should only provide working snippets. Also, are you sure that running in mkShellNoCC will not default to system gcc on non NixOS configurations ?
At this point I am a bit lost in what is the current config, and why. This should probably move to a separate section, and keep the guide focussed on the intended, canonical way to use it with the benefits.

Reply:
Right, I see what you both mean. I'll rework this section with that feedback in mind.

Reply:
I took out the error, and explained that bit in detail in words instead.